contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,262 @@
+        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], _rewardsPercentagePPM, PPM_RESOLUTION);
+        }
+
+        // convert tokens to BNT
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            // skip token if no token balance
+            if (balances[i] == 0) {
+                continue;
+            }
+            // check if token is BNT - we don't need to swap in this case
+            if (token == _bnt) {
+                continue;
+            }
+
+            // get trade amount for token
+            uint256 tradeAmount = balances[i] - rewardAmounts[i];
+
+            // approve tokens for trading on Bancor Network V3
+            _setAllowance(token, tradeAmount);
+
+            uint256 val = token.isNative() ? tradeAmount : 0;
+
+            // swap token to BNT using Bancor Network V3
+            _bancorNetwork.tradeBySourceAmount{ value: val }(token, _bnt, tradeAmount, 1, block.timestamp, address(0));
+        }
+
+        // allocate rewards to caller and burn the rest
+        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    }
+
+    /**
+     * @dev allocates the rewards to caller and burns the rest
+     */
+    function _allocateRewards(address sender, Token[] calldata tokens, uint256[] memory rewardAmounts) private {
+        // transfer the rewards to caller
+        uint256 len = tokens.length;
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            uint256 rewardAmount = rewardAmounts[i];
+            if (rewardAmount == 0) {
+                continue;
+            }
+            if (token.isNative()) {
+                // using a regular transfer here could revert due to exceeding the 2300 gas limit
+                // which is why we're using call instead (via sendValue)
+                payable(sender).sendValue(rewardAmount);
+            } else {
+                token.safeTransfer(sender, rewardAmount);
+            }
+        }
```

Comment:
Maybe do this as part of the loop in function `execute`

Reply:
Hmm, I think it's better to leave it as it is - it becomes difficult to track the logic flow otherwise imo